SNOMED CT REST API improvements

cmark opened the pull request:
This PR adds quality of life improvements to the SNOMED CT REST API:
* Add `sort` query parameter to all search endpoints that have internal support for field-based sorting (`/concepts`, `/descriptions`, `/relationships`, `/refsets`, `/members`, `/classifications`, `/branches`).
* It is possible to specify the order for a given sortKey (`<field>:(asc|desc)`)
* Add `term` based sorting to `concepts` endpoint (`sort=term:desc`)
* Add `parent`, `ancestor`, `statedParent`, `statedAncestor` based filtering directly to `/concepts` endpoint
* Add POST HTTP method variants at `/search` endpoints for `/:path/concepts`, `/:path/descriptions`, `/:path/relationships`, `/:path/members`
* Fix styling of client and server errors section
* Fix nested SNOMED CT Query evaluations
* Fix `ascending` flag bug for script-based sorts

norban committed:
[server] add new package

norban committed:
[rest] support sort key definition

norban committed:
[server] add new helper function

norban committed:
[server] add list to hold all fields

cmark committed:
[api] add sortKeys to all search endpoints that support it

Remove duplicate acceptLanguage header parsing.
Add Fields classes to SNOMED CT domain classes.
Improve sortKey pattern matching (allow only a predefined set of field
names to be recognized as sort field).

cmark committed:
[api] add parent/ancestor/statedParent/statedAncestor filter

cmark committed:
SO-3614 #resolve

Handle nested query expressions properly.

cmark committed:
[docs] fix styling of client and server errors section

cmark committed:
[api] fix bug in AbstractRestService.extractSortFields

cmark committed:
[api] fix sort key pattern in AbstractRestService

cmark committed:
[api] fix ascending bug in SearchResourceRequest.ScriptSort

cmark committed:
[launch] update snow-owl-oss launch config

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/SearchResourceRequest.java:
```diff
@@ -114,6 +123,10 @@ public boolean isAscending() {
 			return ascending;
 		}
 		
+		public static SortScript of(String script, final Map<String, Object> arguments, boolean isAscending) {
```
Please rename the third argument to `ascending`, and call the three-arg `of` factory method in the two-arg `of` methods below as well, for symmetry 🥂

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/domain/SnomedConcept.java:
```diff
@@ -342,6 +371,7 @@ public SnomedConcepts getStatedDescendants() {
 		return statedParentIds == null ? null : Arrays.stream(statedParentIds).mapToObj(Long::toString).collect(Collectors.toList());
 	}
 	
+	@JsonIgnore
```
I'm afraid of situations where someone sends in JSON with an updated enumeration value, and we don't act on it 😨 

Can/should we keep writeability while it is present in the JSON output?

cmark replied:
We use special input types on the REST API side, that will prevent the scenario you are describing.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/entry/SnomedDocument.java:
```diff
@@ -131,10 +132,10 @@ public B effectiveTime(final long effectiveTime) {
-		public static final String RELEASED = "released";
-		public static final String ACTIVE = "active";
-		public static final String EFFECTIVE_TIME = "effectiveTime";
+		public static final String MODULE_ID = SnomedComponent.Fields.MODULE_ID;
```
This introduces coupling between the domain and the indexed document world, which makes future changes to either of the classes harder (either one has to remember to rename properties in both classes, or rename in one and "disconnect" the constant value in the other)

cmark replied:
Then let's connect both to `SnomedRf2Headers` constants, where applicable.

cmark committed:
[api] add POST `/:path/concepts/search` endpoint

cmark committed:
[api] use the same model class for GET and POST concept search

cmark committed:
[api] add POST `/:path/descriptions/search` endpoint

cmark committed:
[api] add POST `/:path/relationships/search` endpoint

cmark committed:
[api] add POST `/:path/members/search` endpoint

cmark committed:
[api] add id filter to all search endpoints

Add languageCode filter to description search endpoints.
Fix naming of members and refsets categories.

cmark committed:
[api] rename `isAscending` to `ascending`

cmark committed:
[api] rename `parents`/`ancestors` to `parent`/`ancestor`

cmark committed:
[index] use constants declared in SnomedRf2Headers in Fields

nagyo reviewed (approved):
LGTM!
